### ISSUE TITLE
ref: update `always_changed` status (due to checkpoint outs)

### ISSUE THREAD

**Issue by shcheklein** [iterative/dvc.org#2428]
Summary: jump to https://github.com/iterative/dvc.org/issues/2428#issuecomment-840227075

Had a discussion with @mattseddon today where we realized:

Seems like the description [here](https://dvc.org/doc/command-reference/status#local-workspace-status) is not exactly correct/complete.

1. `... the stage in dvc.yaml has the always_changed: true` part that refers to stages.
    - _Question:_ what if that stage doesn't have the `always_changed: true` set explicitly but also doesn't have dependencies - is it considered `always_changed`?
    - _Question:_ if I have an implicit "always changed" stage (because of the `checkpoint: true`, or because it doesn't have deps, etc) do we ignore the explicit flag `always_changed: true`? Would it throw an error?

2. With experiments checkpoint outputs make stages "always changed" I believe - we need to clarify and update this as well
   - _Question:_ what if I set the `always_changed: false` flag for this stage?

**Comment by pmrowla** [comment]
The second point is correct, `checkpoint` outputs make a stage `always_changed`

**Comment by shcheklein** [comment]
> The second point is correct, checkpoint outputs make a stage always_changed

then the good question is - is `always_changed` in this case ignored? :)

**Comment by dberenbaum** [comment]
@shcheklein Can you clarify a bit?

>     * `the stage in dvc.yaml has the always_changed: true` - if stage doesn't have dependencies and doesn't have the explicit flag is still considered always changed? (good thing to test in DVC is to try to set the always changed to .dvc and stage w/o dependencies)

What `explicit flag`? I'm also not clear what your statement in parentheses means.

In general, a stage will be considered always changed even if it doesn't have dependencies. For example:

```console
█▓▒░dave@davids-air█▓▒░ Thu Apr 29 08:53:51P
~/repo/ dvc stage add --force -n test --always-changed "echo running stage"             Creating 'dvc.yaml'
Adding stage 'test' in 'dvc.yaml'

To track the changes with git, run:

        git add dvc.yaml
█▓▒░dave@davids-air█▓▒░ Thu Apr 29 08:55:11P
~/repo/ dvc repro
+--------------------------------------------+
|                                            |
|     Update available 2.0.13 -> 2.0.18      |
|       To upgrade follow these steps:       |
|          1. Uninstall dvc binary           |
|          2. Go to https://dvc.org          |
|     3. Download and install new binary     |
|                                            |
+--------------------------------------------+

Running stage 'test':
> echo running stage
running stage
Generating lock file 'dvc.lock'
Updating lock file 'dvc.lock'

To track the changes with git, run:

        git add dvc.lock
Use `dvc push` to send your updates to remote storage.
█▓▒░dave@davids-air█▓▒░ Thu Apr 29 08:55:15P
~/repo/ dvc repro
+--------------------------------------------+
|                                            |
|     Update available 2.0.13 -> 2.0.18      |
|       To upgrade follow these steps:       |
|          1. Uninstall dvc binary           |
|          2. Go to https://dvc.org          |
|     3. Download and install new binary     |
|                                            |
+--------------------------------------------+

Running stage 'test':
> echo running stage
running stage
Use `dvc push` to send your updates to remote storage.
```

> then the good question is - is `always_changed` in this case ignored? :)

Again, could you clarify what you mean here? Checkpoint outputs require stages to be `always_changed` so that every time someone runs that stage it generates a new checkpoint regardless of whether dependencies have changed.

**Comment by shcheklein** [comment]
Sure, I mean the `always_changed` flag here https://dvc.org/doc/user-guide/project-structure/pipelines-files#stage-entries 

> I'm also not clear what your statement in parentheses means.

Yep, sorry about that. I've expanded the description. Please let me know if it makes more sense now.

> Again, could you clarify what you mean here?

same, tried to clarify in the description, PTAL

**Comment by pmrowla** [comment]
> 2. With experiments checkpoint outputs make stages "always changed" I believe - we need to clarify and update this as well
>    
>    * _Question:_ what if I set the `always_changed: false` flag for this stage?

The way we evaluate whether or not a stage is "always changed" is essentially `if (always_changed or checkpoint)`, so the presence of a checkpoint stage will override setting `always_changed: false`

**Comment by jorgeorpinel** [comment]
> if that stage doesn't have the always_changed: true set explicitly but also doesn't have dependencies - is it considered always_changed?

No, `status` considers it up to date as long as it's output(s) exist, in which case `repro` doesn't run it (just tried it).

> if I have an implicit "always changed" stage ... do we ignore the explicit flag always_changed: true? Would it throw an error?
what if I set the always_changed: false flag for this stage?

Specifying any `always_changed` has no effect in that case (per @pmrowla's answer + I tried it). No error is thrown (should it?)

> experiments checkpoint outputs make stages "always changed" I believe - we need to clarify and update this

➕1 on completing the ref with that info, in fact that's the only missing part, I think?

p.s.

> `|     Update available 2.0.13 -> 2.0.18      |`

@dberenbaum looks like you should update DVC 😋

**Comment by jorgeorpinel** [comment]
Related support case: https://discuss.dvc.org/t/experiments-with-checkpoint/744

So to recap here:

- Update `always_changed` description in `dvc status` to mention checkpoint outputs force stages into this state ("so that every time someone runs that stage it generates a new checkpoint regardless of whether dependencies have changed")
- Probably also mention vice versa in some of the checkpoints docs.
Yes?
